vmdb/app/models/metric/ci_mixin/capture/openstack_base.rb

```diff
@@ -37,134 +37,364 @@ def perf_release_openstack
+            metrics_avg[c] = multi_counter_metrics[c][:avg] - last_multi_counter_metrics[c][:avg]
+          end
+          metrics_intervals = {}
+          i[:openstack_counters].each do |c|
+            metrics_intervals[c] = (multi_counter_metrics[c][:duration_end] -
+                                    last_multi_counter_metrics[c][:duration_end])
+          end
+        else
+          metrics_avg = {}
+          i[:openstack_counters].each { |c| metrics_avg[c] = multi_counter_metrics[c][:avg] }
+          # We care about intervals only for diff metrics
+          metrics_intervals = nil
         end
+
+        # If we have found all data samples for all counters, we can compute the value and continue to store it
+        # across the whole <multi_counter_aligned_start, period> interval
+        value = i[:calculation].call(metrics_avg, metrics_intervals)
+      else
+        # Keeping multi_counter_aligned_start and already obtained multi_counter metrics and intervals and thus
+        # expanding the interval for finding all values.
+        next
+      end
+
+      # For (temporary) symmetry with VIM API we create 20-second intervals.
+      (multi_counter_aligned_start + 20.seconds..multi_counter_aligned_end).step_value(20.seconds).each do |ts_item|
+        counter_values_by_ts.store_path(ts_item.iso8601, i[:vim_style_counter_key], value)
+      end
+
+      # Moving multi_counter_aligned_start to next period, this period has been already covered
+      multi_counter_aligned_start       = multi_counter_aligned_end
+      # Moving also guard to next period, will be initialized when next period begins
+      multi_counter_aligned_start_guard = 'initialize_with_period'
+      # Reset multi_counter_aligned_end, which will be initialized to period in next period
+      multi_counter_aligned_end         = nil
+      # Moving completed multi_counter_metrics to last_multi_counter_metrics and nullifying multi_counter_metrics,
+      # so they can be collected for next period.
+      last_multi_counter_metrics        = multi_counter_metrics
+      multi_counter_metrics             = {}
+    end
+  end
+
+  def process_multi_counter_metrics(i, metrics, multi_counter_metrics)
+    # We have to make sure all counters have metric values present.
+    i[:openstack_counters].each do |c|
+      next if metrics.fetch_path(c, :avg).blank?
+
+      # Always overwriting the multicounter metrics. So if new value comes, it will overwrite the old one, that is
+      # possible orphan from another Ceilometer collection period.
+      multi_counter_metrics[c] = metrics[c]
+    end
+    all_multi_counter_metrics_available? i, multi_counter_metrics
+  end
+
+  def all_multi_counter_metrics_available?(i, multi_counter_metrics)
+    all_metrics_available = true
+    i[:openstack_counters].each do |c|
+      if multi_counter_metrics.fetch_path(c, :avg).blank?
+        all_metrics_available = false
       end
     end
+    all_metrics_available
```

Comment:
This can all be simplified to 
`i[:openstack_counters].all? { |c| multi_counter_metrics.fetch_path(c, :avg).present? }`